include/nbl/video/utilities/CComputeBlit.h

```diff
@@ -0,0 +1,900 @@
+	}
+
+	core::smart_refctd_ptr<video::IGPUDescriptorSetLayout> getDSLayout(const uint32_t descriptorCount, const asset::E_DESCRIPTOR_TYPE* descriptorTypes, video::ILogicalDevice* logicalDevice) const
+	{
+		constexpr uint32_t MAX_DESCRIPTOR_COUNT = 5;
+		assert(descriptorCount < MAX_DESCRIPTOR_COUNT);
+
+		video::IGPUDescriptorSetLayout::SBinding bindings[MAX_DESCRIPTOR_COUNT] = {};
+
+		for (uint32_t i = 0u; i < descriptorCount; ++i)
+		{
+			bindings[i].binding = i;
+			bindings[i].count = 1u;
+			bindings[i].stageFlags = asset::IShader::ESS_COMPUTE;
+			bindings[i].type = descriptorTypes[i];
+		}
+
+		auto dsLayout = logicalDevice->createDescriptorSetLayout(bindings, bindings + descriptorCount);
+		return dsLayout;
+	}
+
+	static inline asset::E_FORMAT getCompatClassFormat(const asset::E_FORMAT format)
+	{
+		const asset::E_FORMAT_CLASS formatClass = asset::getFormatClass(format);
+		switch (formatClass)
+		{
+		case asset::EFC_8_BIT:
+			return asset::EF_R8_UINT;
+		case asset::EFC_16_BIT:
+			return asset::EF_R16_UINT;
+		case asset::EFC_32_BIT:
+			return asset::EF_R32_UINT;
+		case asset::EFC_64_BIT:
+			return asset::EF_R32G32_UINT;
+		case asset::EFC_128_BIT:
+			return asset::EF_R32G32B32A32_UINT;
+		default:
+			return asset::EF_UNKNOWN;
+		}
+	}
+
+	//! This calculates the inclusive upper bound on the preload region i.e. it will be reachable for some cases. For the rest it will be bigger
+	//! by a pixel in each dimension.
+	//! Used to size the shared memory.
+	inline core::vectorSIMDu32 getPreloadRegion(
+		const core::vectorSIMDu32& outputTexelsPerWG,
+		const asset::IImage::E_TYPE imageType,
+		const core::vectorSIMDf& scaledMinSupport,
+		const core::vectorSIMDf& scaledMaxSupport,
+		const core::vectorSIMDf& scale)
+	{
+		core::vectorSIMDu32 preloadRegion = core::vectorSIMDu32(core::floor((scaledMaxSupport - scaledMinSupport) + core::vectorSIMDf(outputTexelsPerWG - core::vectorSIMDu32(1, 1, 1, 1)) * scale)) + core::vectorSIMDu32(1, 1, 1, 1);
+
+		// Set the unused dimensions to 1 to avoid weird behaviours with scaled kernels
+		for (auto axis = imageType + 1; axis < 3u; ++axis)
+			preloadRegion[axis] = 1;
+
+		return preloadRegion;
+	}
+
+	//! Query shared memory size for a given `outputTexelsPerWG`.
+	size_t getRequiredSharedMemorySize(
+		const core::vectorSIMDu32& outputTexelsPerWG,
+		const core::vectorSIMDu32& outExtent,
+		const asset::IImage::E_TYPE imageType,
+		const core::vectorSIMDf& scaledMinSupport,
+		const core::vectorSIMDf& scaledMaxSupport,
+		const core::vectorSIMDf& scale,
+		const uint32_t channelCount)
+	{
+		const auto preloadRegion = getPreloadRegion(outputTexelsPerWG, imageType, scaledMinSupport, scaledMaxSupport, scale);
+
+		const size_t requiredSmem = ((preloadRegion.x * preloadRegion.y * preloadRegion.z) + core::max(outputTexelsPerWG.x * preloadRegion.y * preloadRegion.z, outputTexelsPerWG.x * outputTexelsPerWG.y * preloadRegion.z)) * channelCount * sizeof(float);
```

Comment:
you cannot alias X-axis pass output scratch with Y-axis output scratch, the `max` is in the wrong place

Reply:
Apologies, old code.

Resolved, now.